### PR TITLE
[Fix #5217] CLEANUP - Expand ProcessedSource public API

### LINKED ISSUE #5217: Encapsulate ProcessedSource & Token related logic in new Module
In an effort to move non-cop related logic out of cop & cop mixin implementation code, I think it would valuable to extend `ProcessedSource` with helper methods. However, right now our `ProcessedSource` file mostly contains parser related logic, and it looks like it would be more readable/eloquent to separate `processed_source` logic directly related to tokens into a new Module. Perhaps `ProcessedTokens`, `ProcessedTokenLogic` (like `AutocorrectLogic`), or just maybe `ProcessedSourceHelpers`?

Here is an example.  [`Layout/ExtraSpacing`'s `aligned_comments?(token)` implementation](https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/layout/extra_spacing.rb#L158-L176). Here are those lines:

```ruby
def aligned_comments?(token)
  ix = processed_source.comments.index do |c|
    c.loc.expression.begin_pos == token.begin_pos
  end
  aligned_with_previous_comment?(ix) || aligned_with_next_comment?(ix)
end

def aligned_with_previous_comment?(ix)
  ix > 0 && comment_column(ix - 1) == comment_column(ix)
end

def aligned_with_next_comment?(ix)
  ix < processed_source.comments.length - 1 &&
    comment_column(ix + 1) == comment_column(ix)
end

def comment_column(ix)
  processed_source.comments[ix].loc.column
end
```

The first method in this group is the only publicly consumable method outside the scope of this collection. The rest of these methods could be modified as private. 

This is just one example, but I am confident we have a lot of `processed_source` logic that deals with passing tokens, token lines/columns, and all sorts of other computations for alignment, spacing, and indentation that could refactored to simplify our cop code. (I've written a chunk of it.) WDTY?

### PR DESCRIPTION
**Pull request by garettarrowood**
This PR adds custom enumerables, predicates, delegators and helper methods to `ProcessedSource`, and adds tests for each addition. Overall, these changes reduce method chaining and provide a clearer API. They also relocate many cop/mixin methods into a `ProcessSourceLogic` module. These methods operate solely on a`processed_source` and thus were the responsibility of `ProcessedSource` and not the concern of a cop/mixin.

See comments for specifics.

-----------------

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/correctors/alignment_corrector.rb [hunk 1]
```diff
@@ -96,10 +96,6 @@ def each_line(expr)
           end
         end
 
-        def begins_its_line?(range)
-          (range.source_line =~ /\S/) == range.column
-        end
-
```
Comment (garettarrowood) [inline lib/rubocop/cop/correctors/alignment_corrector.rb]: Unrelated cleanup. It looks like during the Corrector class breakout, this method got duplicated. It already exists in `util.rb` which is included in this class.

File: lib/rubocop/cop/mixin/preceding_following_alignment.rb [hunk 2]
```diff
@@ -55,7 +55,7 @@ def aligned_with_line?(line_nos, range, indent = nil)
         false
       end
 
-      def comment_lines
```
Comment (garettarrowood) [inline lib/rubocop/cop/mixin/preceding_following_alignment.rb]: I ran into another method called `comment_lines` that is moved to the new module. I thought this may be a place to swap in the new method, but this `comment_lines` takes alignment into account. So I gave it more descriptive name.

File: lib/rubocop/cop/processed_source_logic.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # This module extends ProcessedSource with helper functionality.
+    module ProcessedSourceLogic
```
Comment (garettarrowood) [inline lib/rubocop/cop/processed_source_logic.rb]: I can't decide whether it nice to separate this logic into a module, or to just include it in `ProcessedSource` directly. Also, I am open to suggestions if someone feels strongly about a different name.

Comment (bbatsov) [reply lib/rubocop/cop/processed_source_logic.rb]: A mixin that's extremely class-specifix is not really a mixin IMO. I'd put everything in `ProcessedSource` probably (at least pretty common methods). I see here and there methods that seem like something that's used just once or twice and I wonder what to do with them. Perhaps its fine for such cases to just keep this logic in the relevant cops.

Comment (garettarrowood) [reply lib/rubocop/cop/processed_source_logic.rb]: > A mixin that's extremely class-specific is not really a mixin IMO.

Agreed. My reasoning for going this route was the way we include `AutocorrectLogic` into our `Cop` class.  In a separate PR, I can move these methods directly into `cop.rb`.

> I'd put everything in ProcessedSource probably (at least pretty common methods). I see here and there methods that seem like something that's used just once or twice and I wonder what to do with them. Perhaps its fine for such cases to just keep this logic in the relevant cops.

I'm torn about this. 

I think the one I listed in the issue, `aligned_comments?`, is the worst example here. It seems scoped to alignment issues and thus does seem best placed back into its alignment mixin. `empty_brackets?` looks like it should be moved back as well. These two methods required very specific args. One has to be a comment token, and the latter has to be opening and closing bracket tokens. Reuse seems unlikely.

But the rest of them (the ones currently in the Module), although only used once or twice, seem more like the responsible of a `processed_source` and not a cop/mixin. The args for these are more generic (a token, a range, an index, a line, a string).

I'll make these alterations. Feel free to nudge me again to move more back :) .

Comment (garettarrowood) [reply lib/rubocop/cop/processed_source_logic.rb]: Hmm.  `previous_and_current_lines_empty?` and `stripped_upto?` also seem to have a narrow use case. I'll move these back too.

File: lib/rubocop/cop/mixin/def_node.rb [hunk 4]
```diff
@@ -21,6 +21,10 @@ def preceding_non_public_modifier?(node)
         end
       end
 
+      def stripped_source_upto(index)
+        processed_source[0..index].map(&:strip)
+      end
+
```
Comment (garettarrowood) [inline lib/rubocop/cop/mixin/def_node.rb]: This was originally in `util.rb`. But this mixin is the only place the method is used. So here it is.

**Comment by bbatsov** [comment]
There's now one merge conflict that needs to be addressed.

**Comment by bbatsov** [comment]
Once again - a big thanks from me for taking so much time to improve the internals of RuboCop! 🙇

**Comment by garettarrowood** [comment]
You're welcome! Thanks for actively maintaining this great repo!  It's a pleasure hacking on the codebase.

**Comment by jonas054** [comment]
@garettarrowood Great job! One minor thing: when there are no user-observable changes, you shouldn't update `CHANGELOG.md`.
